docs/SPIR-V.rst

```diff
@@ -0,0 +1,194 @@
+==============
+SPIR-V Codegen
+==============
+
+.. contents::
+   :local:
+   :depth: 2
+
+Introduction
+============
+
+This document describes the designs and logistics for supporting SPIR-V codegen functionality. `SPIR-V <https://www.khronos.org/registry/spir-v/>`_ is a binary intermediate language for representing graphical-shader stages and compute kernels for multiple Khronos APIs, such as Vulkan, OpenGL, and OpenCL. At the moment we only intend to support the Vulkan favor of SPIR-V.
+
+DirectXShaderCompiler is the reference compiler for HLSL. Adding SPIR-V codegen in DirectXShaderCompiler will enable the usage of HLSL as a frontend language for Vulkan shader programming. Sharing the same code base also means we can track the evolution of HLSL more closely and always deliver the best of HLSL to developers. Moreover, developers will also have a unified compiler toolchain for targeting both DirectX and Vulkan. We believe this effort will benefit the general graphics ecosystem.
+
+Designs
+=======
+
+Various designs are driven by technical considerations together with the following guidelines for good citizenship within DirectXShaderCompiler:
+
+- Conduct minimal changes to existing interfaces and libraries
+- Perfer less intrusive solutions
+
+General approach
+----------------
+
+The general approach is to translate frontend AST directly into SPIR-V binary. We choose this approach considering that
+
+- Frontend AST is much more higher-level than DXIL. For example, `DXIL scalarized vectors <https://github.com/Microsoft/DirectXShaderCompiler/blob/master/docs/DXIL.rst#vectors>`_ but SPIR-V has native support.
+- DXIL has widely different semantics than Vulkan favor of SPIR-V. For example, `structured control flow is not preserved in DXIL <https://github.com/Microsoft/DirectXShaderCompiler/blob/master/docs/DXIL.rst#control-flow-restrictions>`_ but SPIR-V requires it.
+- Frontend AST perserves the information in the source code better.
```

Comment:
Thanks for adding.  

Example: If there is a feature that Vulkan does not support, but is allowed by HLSL (for DX12), then we _should_ generate an error message saying so, rather than proceeding to generate SPIR-V that no driver would accept.